dominion/src/dominion/arena.rs

```diff
@@ -115,88 +119,103 @@ impl Arena {
+                    *total_wealth += gains.worth;
+
+                    for _ in 0..gains.card {
+                        self.players[player_id].draw_card();
+                    }
+
+                    Ok(())
+                }
+            }
+            TurnPhase::Buy { .. } => Err(Error::WrongTurnPhase),
+        }
+    }
+
+    pub fn play_treasure(&mut self, player_id: usize, card: CardKind) -> Result<()> {
+        self.check_active(player_id)?;
+
+        let card_index = self.players[player_id]
+            .hand
+            .iter()
+            .position(|&x| x == card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        let gains = card.treasure().ok_or_else(|| Error::InvalidCardChoice)?;
+
+        match &mut self.turn.phase {
+            TurnPhase::Action { .. } => Err(Error::WrongTurnPhase),
             TurnPhase::Buy {
                 remaining_buys,
                 total_wealth,
-            } => match location {
-                Location::Hand => match self.players[player_id].hand.remove_item(&card) {
-                    Some(card) => match card.treasure() {
-                        Some(i) => {
-                            *total_wealth += i;
-                            self.players[player_id].play_zone.push(card);
-                            Ok(())
-                        }
-                        None => {
-                            self.players[player_id].hand.push(card);
-                            Err(Error::InvalidCardChoice)
-                        }
-                    },
-                    None => Err(Error::InvalidCardChoice),
-                },
-                Location::Supply => match self.supply.get_mut(card) {
-                    Some(supply_count) => {
-                        if *supply_count == 0 {
-                            Err(Error::NoMoreCards)
-                        } else if *remaining_buys == 0 {
-                            Err(Error::NoMoreBuys)
-                        } else if *total_wealth < card.cost() {
-                            Err(Error::NotEnoughWealth)
-                        } else {
-                            *supply_count -= 1;
-                            *remaining_buys -= 1;
-                            *total_wealth -= card.cost();
-                            self.players[player_id].discard_pile.push(card);
-                            Ok(())
-                        }
-                    }
-                    None => Err(Error::InvalidCardChoice),
-                },
-                _ => Err(Error::InvalidCardLocation),
-            },
+            } => {
+                let player = &mut self.players[player_id];
+                let card = player.hand.remove(card_index);
+                player.play_zone.push(card);
+
+                *total_wealth += gains;
+
+                Ok(())
+            }
+        }
+    }
+
+    pub fn buy_card(&mut self, player_id: usize, card: CardKind) -> Result<()> {
```

Comment:
Why would we ever want to call this with a player_id != current player id? It doesn't even work because of the assert in that case. Either remove the assertion or the argument.

Reply:
Removing the assertion means that we'll be using the wrong turn phase. So remove the player_id parameter.

Reply:
We'll need to validate a whole bunch of things because we can't trust the input. This does not protect against another player trying to force the current player to buy a curse card. 

The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

Reply:
> The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

I agree this makes sense and can see that the function `play_action`/`buy_card`/etc. would be simpler without checking ` player_id`. However, I don't want remove `player_id` as a parameter until the privilege checking is implemented and I want to test out my effect stack idea first =p

Reply:
Then debug_assert!(...) is more appropriate